pkg/controller/argocd/role.go

```diff
@@ -83,30 +85,54 @@ func (r *ReconcileArgoCD) reconcileRoles(cr *argoprojv1a1.ArgoCD) (role *v1.Role
 }
 
 // reconcileRole
-func (r *ReconcileArgoCD) reconcileRole(name string, policyRules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD) (*v1.Role, error) {
-	role := newRole(name, policyRules, cr)
-	if err := applyReconcilerHook(cr, role, ""); err != nil {
+func (r *ReconcileArgoCD) reconcileRole(name string, policyRules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD) ([]*v1.Role, error) {
```

Comment:
I think we really need some documentation within this method. Why is any role created, or not. This is a critical path function. Please help others who need to modify this in the future to understand _why_ something is done the way it is done. Don't assume any context knowledge.

Reply:
`reconcileRole` returns a list of roles for all namespaces which is managed by a single ArgoCD instance. So for example if `foo and bar` is managed by an instance of ArgoCD, then it would return a list of roles for `application-controller/argocd-server ... etc` for each namespace. 
This list of roles is then used by `reconcileRolebinding` to create roleBindings for each role. 

Although i see this as an issue, this would create issues while trying to remove the rolebindings say for `DEX` if it gets disabled.

Reply:
Added comments!